common/src/main/java/com/proximyst/ban/commands/HistoryCommand.java

```diff
@@ -63,17 +65,31 @@ public void register(final @NonNull CommandManager<@NonNull IBanAudience> comman
   private void execute(final @NonNull CommandContext<IBanAudience> ctx) {
     final BanUser target = ctx.get("target");
 
-    this.messageService.sendFormattedMessage(ctx.getSender(), Identity.nil(), MessageKey.COMMANDS_FEEDBACK_HISTORY,
-        "targetName", target.getUsername(),
-        "targetUuid", target.getUuid())
-        .exceptionally(this.banExceptionalFutureLogger.cast());
+    this.messageService.commandsFeedbackHistory(target)
+        .send(ctx.getSender());
 
     this.punishmentService.getPunishments(target.getUuid())
-        .thenCompose(punishments -> this.messageService.formatHistory(punishments, target))
-        .thenAccept(messages -> {
-          for (final Component message : messages) {
-            ctx.getSender().sendMessage(Identity.nil(), message);
+        .thenCompose(punishments -> {
+          final CompletableFuture<?>[] futures = new CompletableFuture<?>[punishments.size()];
+          int index = 0;
+          for (final Punishment punishment : punishments) {
+            futures[index++] = this.messageService.commandsHistoryEntry(punishment)
+                .component();
```

Comment:
This loop is a little too funky for me. A standard for loop or a Stream would work better.

Reply:
Both of the suggestions would be wasteful, I'm afraid.